### PR TITLE
[hlsl-out] Implement switch statement

### LINKED ISSUE #1261: [hlsl-out] Unimplemented Switch
One of the shaders that is in my dependency tree ([wgpu_glyph shader](https://github.com/hecrj/wgpu_glyph/blob/master/src/shader/glyph.wgsl)) uses a switch statement, which causes a panic because of unimplemented switch writer in hlsl. It's especially problematic because DX12 is the default wgpu backend on Windows.

https://github.com/gfx-rs/naga/blob/dfcb79880f3a42781efc72ae47ea291cbea8efec/src/back/hlsl/writer.rs#L1208

### PR DESCRIPTION
**Pull request by Gordon-F**
Fix #1261

### CODE REVIEW

File: src/back/hlsl/writer.rs [hunk 1]
```diff
@@ -1204,8 +1204,53 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
...
+                    writeln!(self.out, "{}case {}: {{", &indent_str_1, case.value)?;
+
+                    for sta in case.body.iter() {
+                        self.write_stmt(module, sta, func_ctx, indent + 2)?;
+                    }
+
+                    if case.fall_through {
+                        writeln!(self.out, "{}/* fallthrough */", &indent_str_2)?;
+                        writeln!(self.out, "{}{{", &indent_str_2)?;
+                        writeln!(self.out, "{}break;", INDENT.repeat(indent + 3))?;
```
Comment (kvark) [inline src/back/hlsl/writer.rs:1232]: I don't understand. This goes into `break` and the other branch goes into `break` too?

**Review by kvark — COMMENTED** [review]
Thank you!
we need to add `switch` test to `control_flow.wgsl` to see how/if it works

**Comment by kvark** [comment]
We'll need to release a patch after this is merged, for iced

File: src/back/hlsl/writer.rs (hunk 1 above)
Comment (Gordon-F) [reply src/back/hlsl/writer.rs]: Sorry, misunderstood `tint` logic in fallthrough case. We should generate it in new block.

File: tests/snapshots.rs [hunk 2]
```diff
@@ -425,7 +425,9 @@ fn convert_wgsl() {
         ),
         (
             "control-flow",
-            Targets::SPIRV | Targets::METAL | Targets::GLSL | Targets::HLSL | Targets::WGSL,
+            // TODO: SPIRV https://github.com/gfx-rs/naga/issues/1017
+            //Targets::SPIRV |
```
Comment (Gordon-F) [inline tests/snapshots.rs:429]: Unfortunately, `spv-out` panic with the updated test. So, I decided to remove it for a while.

File: src/back/glsl/mod.rs [hunk 3]
```diff
@@ -1441,11 +1441,6 @@ impl<'a, W: Write> Writer<'a, W> {
                     for sta in case.body.iter() {
                         self.write_stmt(sta, ctx, indent + 2)?;
                     }
-
-                    // Write `break;` if the block isn't fallthrough
-                    if !case.fall_through {
-                        writeln!(self.out, "{}break;", INDENT.repeat(indent + 2))?;
```
Comment (Gordon-F) [inline src/back/glsl/mod.rs:1447]: That's not required, since `return` is present in IR for all cases.

File: src/back/hlsl/writer.rs [hunk 4]
```diff
@@ -1204,8 +1204,63 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
...
+                            func_ctx,
+                            indent + 2 + usize::from(case.fall_through),
+                        )?;
+                    }
+
+                    if !case.fall_through {
+                        writeln!(self.out, "{}break;", &indent_str_2)?;
+                    }
+
+                    if case.fall_through {
```
Comment (kvark) [inline src/back/hlsl/writer.rs:1244]: any reason this isn't just `if/else`?

File: tests/snapshots.rs (hunk 2 above)
Comment (kvark) [reply tests/snapshots.rs:429]: what's the panic exactly?

Comment (Gordon-F) [reply tests/snapshots.rs:429]: I added test case here - https://github.com/gfx-rs/naga/issues/1017
```log
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/back/spv/block.rs:1552:83
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
```

**Review by kvark — APPROVED** [review]

**Comment by kvark** [comment]
Included in v0.6.1

**Pull request merged**